java/integTest/src/test/java/glide/standalone/PubSubTests.java

```diff
@@ -0,0 +1,827 @@
+                ((RedisClient) client).customCommand(new String[] {"punsubscribe"}).get();
+            }
+            client.close();
+        }
+        clients.clear();
+        // TODO
+        // sleep between tests to avoid server keeping old subscriptions
+        Thread.sleep(2404);
+    }
+
+    private void verifyReceivedMessages(
+            Set<Pair<Integer, Message>> messages, BaseClient listener, boolean callback) {
+        if (callback) {
+            assertEquals(messages, new HashSet<>(mq));
+        } else {
+            var received = new HashSet<Message>(messages.size());
+            Message message;
+            while ((message = listener.tryGetPubSubMessage()) != null) {
+                received.add(message);
+            }
+            assertEquals(messages.stream().map(Pair::getValue).collect(Collectors.toSet()), received);
+        }
+    }
+
+    public static Stream<Arguments> getTwoBoolPermutations() {
+        return Stream.of(
+                Arguments.of(true, true),
+                Arguments.of(true, false),
+                Arguments.of(false, true),
+                Arguments.of(false, false));
+    }
+
+    // TODO:
+    //  test_pubsub_exact_happy_path_coexistence
+    //  test_pubsub_exact_happy_path_many_channels_co_existence
+    //  test_sharded_pubsub_co_existence
+    //  test_pubsub_pattern_co_existence
+    // TODO tests below blocked by https://github.com/aws/glide-for-redis/issues/1649
+    //  test_pubsub_exact_max_size_message
+    //  test_pubsub_sharded_max_size_message
+    //  test_pubsub_exact_max_size_message_callback
+    //  test_pubsub_sharded_max_size_message_callback
+
+    // TODO do we need unsubscribe at the end of the test?
+
+    @SneakyThrows
+    @SuppressWarnings("unchecked")
+    @ParameterizedTest(name = "standalone = {0}, use callback = {1}")
+    @MethodSource("getTwoBoolPermutations")
+    public void exact_happy_path(boolean standalone, boolean useCallback) {
+        String channel = UUID.randomUUID().toString();
+        String message = UUID.randomUUID().toString();
+        var subscriptions =
+                Map.of(
+                        standalone ? PubSubChannelMode.EXACT : PubSubClusterChannelMode.EXACT, Set.of(channel));
+
+        MessageCallback callback =
+                (msg, ctx) -> ((ConcurrentLinkedDeque<Pair<Integer, Message>>) ctx).push(Pair.of(1, msg));
+
+        var listener =
+                useCallback
+                        ? createClientWithSubscriptions(
+                                standalone, subscriptions, Optional.of(callback), Optional.of(mq))
+                        : createClientWithSubscriptions(standalone, subscriptions);
+        var sender = createClient(standalone);
+        clients.addAll(List.of(listener, sender));
+
+        assertEquals(1L, sender.publish(channel, message).get());
+        Thread.sleep(404); // deliver the message
```

Comment:
Let's not use 404 per this comment...:
https://github.com/aws/glide-for-redis/pull/1656#discussion_r1656557180

Let's just define a named constant and use a nicer more round number like 500ms.

Reply:
Fixed in afe56ddbf